ipaclient/plugins/cert.py

```diff
@@ -27,17 +32,109 @@
+        database = options.pop('database', None)
+        private_key = options.pop('private_key', None)
+        csr_profile_id = options.pop('csr_profile_id', None)
+        password_file = options.pop('password_file', None)
+
+        if csr is None:
+            if database:
+                helper = u'certutil'
+                helper_args = ['-d', database]
+                if password_file:
+                    helper_args += ['-f', password_file]
+            elif private_key:
+                helper = u'openssl'
+                helper_args = [private_key]
+                if password_file:
+                    helper_args += ['-passin', 'file:%s' % password_file]
+            else:
+                raise errors.InvocationError(
+                    message=u"One of 'database' or 'private_key' is required")
+
+            with tempfile.NamedTemporaryFile(
+                    ) as scriptfile, tempfile.NamedTemporaryFile() as csrfile:
+                # If csr_profile_id is passed, that takes precedence.
+                # Otherwise, use profile_id. If neither are passed, the default
+                # in cert_get_requestdata will be used.
+                profile_id = csr_profile_id
+                if profile_id is None:
+                    profile_id = options.get('profile_id')
+
+                self.api.Command.cert_get_requestdata(
+                    profile_id=profile_id,
+                    principal=options.get('principal'),
+                    out=unicode(scriptfile.name),
+                    helper=helper)
+
+                helper_cmd = [
+                    'bash', '-e', scriptfile.name, csrfile.name] + helper_args
+
+                try:
+                    subprocess.check_output(helper_cmd)
+                except subprocess.CalledProcessError as e:
+                    raise errors.CertificateOperationError(
+                        error=(
+                            _('Error running "%(cmd)s" to generate CSR:'
+                              ' %(err)s') %
+                            {'cmd': ' '.join(helper_cmd), 'err': e.output}))
+
+                try:
+                    csr = unicode(csrfile.read())
+                except IOError as e:
+                    raise errors.CertificateOperationError(
+                        error=(_('Unable to read generated CSR file: %(err)s')
+                               % {'err': e}))
+                if not csr:
+                    raise errors.CertificateOperationError(
+                        error=(_('Generated CSR was empty')))
+
+                return super(cert_request, self).forward(csr, **options)
+        else:
+            if database is not None or private_key is not None:
+                raise errors.MutuallyExclusiveError(reason=_(
+                    "Options 'database' and 'private_key' are not compatible"
+                    " with 'csr'"))
+            return super(cert_request, self).forward(csr, **options)
```

Comment:
Nitpick: if you move this call one intendation level to the left, you will be able to remove the forward call above.